runtime/vm/jvminit.c

```diff
@@ -651,6 +655,29 @@ freeJavaVM(J9JavaVM * vm)
 		pool_kill(vm->cifArgumentTypesCache);
 		vm->cifArgumentTypesCache = NULL;
 	}
+
+	/* Clean up all resources created via allocateUpcallThunkMemory, including the generatered thunk
+	 * and the corresponding metadata of each entry stored in the hashtable.
+	 * See UpcallThunkMem.cpp for details.
+	 */
+	if (NULL != vm->thunkHeapWrapper) {
+		J9UpcallThunkHeapWrapper *thunkHeapWrapper = vm->thunkHeapWrapper;
```

Comment:
looks similar to the code in `Java_jdk_internal_foreign_abi_UpcallStubs_freeUpcallStub0`, can a common helper be used?

Reply:
Functionally speaking, they are different because the code in `Java_jdk_internal_foreign_abi_UpcallStubs_freeUpcallStub0` is to release the specified thunk plus the corresponding metadata of a entry existing in the hashtable when the session/scope is terminated, while the code here is to clean up the whole thunk hashtable by invoking `freeUpcallMetaDataDoFn()` for each entry if there are still thunks left to be released when VM exits. So sharing the code of`freeUpcallMetaDataDoFn()` for a given entry might be possible but unnecessary at this point.